Fix free-memory write in k_thread_abort(_current) from an ISR

andyross opened the pull request:
Discussion in the big saga in #68262 turned up a longstanding issue with the way thread abort works in ISRs: the thread struct itself gets cleaned up, but the exit code on the ISR path still needs to use that _current pointer to do things like save the full pickled context of the interrupted thread in switch_handle (and other arch-specific data too, details vary).  But that thread is dead now, and the caller of `k_thread_abort()` was promised that the memory is free and available for reuse.

This hasn't traditionally been a problem because "most" threads are static and will only be reused for another thread, and we've always disallowed k_thread_create() in ISRs.  But that's not documented, and on SMP this becomes an unresolvable race: other CPUs need to know whether the foreign ISR has completed before they can know it's safe to touch the thread memory, and we don't provide any hooks for that.  (This race was in fact how this was discovered).

The fix is pretty simple though: just freeload on the pre-existing dummy thread handling used at CPU startup to swap _current for a dummy.

I was worried that existing architectures were doing bad magic of their own (e.g. caching a _current pointer in a register), but that seems not to be the case.  The augmented test added to catch post-abort modifications to the thread struct should help to keep the arch layers clean in the future.

The only gotcha currently is POSIX, which unfortunately does a "deferred" abort at ISR exit based on the _current pointer, and isn't expecting that to change.  That's going to need some modification to work with this scheme, so for now it remains susceptible to the same kind of application bug.  But as long as no other platforms allow it, we should be safe enough.

peter-mitsis committed:
tests: thread abort deadlock scenario

Adds a test to verify that a series of k_thread_abort() issued from
ISRs do not cause a deadlock.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

andyross committed:
kernel/sched: Fix free-memory write when ISRs abort _current

After a k_thread_abort(), the resulting thread struct is documented as
unused/free memory that may be re-used (for example, to respawn a new
thread).

But in the special case of aborting the current thread from within an
ISR, that wasn't quite happening.  The scheduler cleanup would
complete, but the architecture layer would still try to context switch
away from the aborted thread on exit, and that can include writes to
the now-reused thread struct!  The specifics will depend on
architecture (some do a full context save on entry, most don't), but
in the case of USE_SWITCH=y it will at the very least write the
switch_handle field.

Fix this simply, with a per-cpu "switch dummy" thread struct for use
as a target for context switches like this.  There is some non-trivial
memory cost to that; thread structs on many architectures are large.

Pleasingly, this also addresses a known deadlock on SMP: because the
"spin in ISR" step now happens as the very last stage of
k_thread_abort() handling, the existing scheduler lock works to
serialize calls such that it's impossible for a cycle of threads to
independently decide to spin on each other: at least one will see
itself as "already aborting" and break the cycle.

Fixes #64646

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
tests/kernel/threads: Augment abort_from_isr test to detect FMW

There's a easily-tripped-upon free memory write condition in the arch
layers where they will write to a cached _current pointer after that
thread has been aborted.  Detect this by clobbering the thread data
after the return from k_thread_abort() and validating that it's still
clear after the ISR returns.

Note that the clobbering of the thread struct requires the removal of
a k_thread_join() that (obviously) requires that it see a DEAD flag
set.  Joining aborted threads isn't actually legal, but does still
work as long as app code doesn't reuse the memory.  Basically we can't
test both cases here, and we have join coverage elsewhere already.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
kernel/sched: Refeactor/cleanup z_thread_halt()

Big change is to factor out a thread_halt_spin() utility to manage the
core complexity of this code: the situation where an ISR is asked to
abort a thread already running on another SMP CPU.

With that gone, things can be cleaned up quite a bit.  Remove early
returns, most of the "#if CONFIG_SMP" usage was superfluous and will
optimize out, unify and clean up the comments, etc...

No behavioral changes (hopefully), just refactoring.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
kernel/spinlock: Fix SPIN_VALIDATE in ISRs

Spinlocks taken in ISRs were storing the _current thread pointer of
the interrupted thread as the owner, which was never strictly correct
but was benign as the thread would never run until the lock was
released.

But now k_thread_abort(_current) in an ISR has been fixed to eliminate
all references to the (now aborted) thread struct, and _current points
to a dummy thread.  Handle that edge case in the validation framework.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
tests/kernel: Bump kobj thread bitmask size for a few tests

A scheduler fix for free memory usage on aborted threads is now using
a per-CPU dummy thread instead of a single stack-based one at startup.
These static thread objects need spots in the kobj bitmasks, and a few
tests are sitting right at the default limit (16 threads).

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
FWIW, this is actually based on the same refactor from the SMP abort proposal too, but doesn't contain it.  It's really best thought of as a separate bug.  But once this is merged, the SMP race becomes trivial to fix.

andyross commented on kernel/sched.c:
```diff
@@ -447,84 +450,69 @@ void z_sched_start(struct k_thread *thread)
+			    is_aborting(_current) ? _THREAD_DEAD : _THREAD_SUSPENDED);
+	}
+	k_spin_unlock(&_sched_spinlock, key);
+	while (is_halting(thread)) {
 	}
+	key = k_spin_lock(&_sched_spinlock);
```
That patch isn't (or shouldn't be) a change to behavior, it just moves the loop into a separate function.  The code in tree right now spins on is_halting() at line 499.  This is the kind of refactoring change that is a lot easier to review on github if you view it in "split" mode.

peter-mitsis reviewed (dismissed):
The whole set of proposed changes is a definite move in the right direction and an improvement to what we currently have. (Incidentally, I like this better than the dummy thread version I was working on.)

As far as I can tell, this also fixes the deadlock issue I described in #64646. As such I am giving this PR a +1 as is. The only concern, and this is something that can be addressed either in a separate PR or as an addendum to this one, is the `is_halting(thread)` spinning previously described.

andyross commented:
> As far as I can tell, this also fixes the deadlock issue I described in https://github.com/zephyrproject-rtos/zephyr/issues/64646.

Not quite, it still has the cycle deadlock where N threads can point at each other and get stuck.  But that's now trivially fixable by a very slightly rebased version of this patch: https://github.com/zephyrproject-rtos/zephyr/pull/70455/commits/727c5606a9d1219621c6b49b32cc6035515292cb  Basically this fixes the hard part where we had to think of a way to whitebox the call to z_sched_switch_spin() by removing the need for that call entirely.

peter-mitsis commented:
I find myself in an unusual position as a reviewer--in opposition to the author but arguing that the proposed code is even better than the author thinks. :)  I've run through my analysis of events a few times with different number of CPUs and each time I find that your current code resolves the deadlock.

At the risk of sharing a dense and tedious analysis, what I find is ...

1. CPU0 in ISR0 and CPU0->current = T0, CPU1 in ISR1 and CPU1->current = T1.
2. CPU0 calls k_thread_abort(T1), CPU1 calls k_thread_abort(T0)
3. CPU0 wins sched_spinlock and marks T1 as aborting
4. CPU0 calls halt_thread_spin(), unlocks sched_spinlock and spins on T1 waiting for it to halt
5. CPU1 wins sched_spinlock and marks T0 as aborting
6. CPU1 calls halt_thread_spin()
6.1. CPU1 halts CPU1->current (aka T1) thereby breaking CPU0 out of its loop
6.2. CPU1->current becomes a dummy thread
6.3. CPU1 unlocks sched_spinlock and spins T0 waiting for it to halt
7. CPU0 wins sched_spinlock, and then unlocks it after it returns from halt_thread_spin()
8. CPU0 finishes k_thread_abort(T1) and exits the ISR
9. CPU0 as part of leaving the ISR calls next_up() and finds that CPU0->current (aka T0) is halting
10. CPU0 halts T0 (breaking CPU1 out of its loop) and sets CPU0->current to a dummy thread
11. CPU0 finishes exiting the ISR and a new thread is scheduled
12. CPU1 wins the sched_spinlock and then unlocks it after returning from halt_thread_spin()
13. CPU1 finishes k_thread_abort(T0) and exits the ISR
14. CPU1 calls next_up(), but CPU1->current is already a dummy thread
15. CPU1 finishes exiting the ISR and a new thread is scheduled

No deadlock.

If you think I missed something I would love to hear it.

cfriedt reviewed (dismissed):
LGTM

andyross commented:
OK, cleanup is pushed, please re-review

peter-mitsis commented on kernel/sched.c:
```diff
@@ -449,19 +454,17 @@ void z_sched_start(struct k_thread *thread)
-	z_sched_switch_spin(thread);
-	return key;
 }
 
```
The new version of thread_halt_spin() is unfortunately wrong as sched_spinlock is no longer being manipulated by this routine. Not unlocking (or releasing) it before the `while (is_halting(thread))` loop prevents another CPU from performing the abort/suspend the target thread (and also leaves the `key` argument unused).

andyross replied:
Oh, for... yeah, I pushed a bad tree.  Looks like I dropped a git add or something and that one line is still sitting in my tree uncommited.  Also it looks like I had an incorrect cut/paste when redoing the commit message of the main patch and it ended up being a copy of the test patch instead.  One sec...

andyross commented:
Needed a rebase too

peter-mitsis commented on kernel/sched.c:
```diff
@@ -454,84 +457,72 @@ void z_sched_start(struct k_thread *thread)
-						  : _THREAD_SUSPENDED);
+			    is_aborting(_current) ? _THREAD_DEAD : _THREAD_SUSPENDED);
 	}
+	k_spin_unlock(&_sched_spinlock, key);
+	while (is_halting(thread)) {
+		arch_spin_relax();
+	}
+}
```
According to the weak definition of arch_spin_relax() in idle.c, this is intended to be called while interrupts are locked (I've been bitten by this before).

andyross replied:
Oh, right.  @npitre needed that IIRC because the relax on IIRC arm64 needs to poll for an IPI because of the somewhat frustrating FPU shootdown deadlock (not unlike this one).  This may need to be a hand-coded relaxation.  I realize I'm not hitting it in CI because your patch with the test got dropped a while back, should just add that to this PR.

peter-mitsis replied:
It may look a little clunky, but we could do ..

```
k_spin_release(&_sched_spinlock);
while (is_halting(thread)) {
    arch_spin_relax();

    arch_irq_unlock(key.key);  /* Unlock/lock interrupts for better */
    key.key = arch_irq_lock();  /* nested interrupt responsiveness */
}
arch_irq_unlock(key.key);
```

andyross replied:
Yeah, that's likely the best way to do it, given that we really do want architectures to be able to define relaxation code on their own (we actually don't need the arm64 magic at all here though, since the released interrupts would allow the FPU IPI to be handled).

andyross commented:
Rebase, fix the interrupt/relax issue, also steal @peter-mitsis 's test patch from the earlier PR which really should have been in this one to start with.

npitre commented:
Quick question: is it really necessary to have a dummy thread struct per CPU?

andyross commented:
> is it really necessary to have a dummy thread struct per CPU?

Sort of.  The way the framework works right now is that everyone in the world gets to assume there's a _current that points to a valid struct.  And this is a circumstance where that can't be an actual thread, because the preempted thread got  aborted and its memory is free to be used for other purposes.  So we need something.

It's nontrivially expensive, it's true.  Though if it helps: note that this now replaces the on-stack dummy thread used at CPU startup, reducing initialization pressure on the interrupt stack[1].  So some apps might be able to shrink that and pay less cost.

The asymptotically ideal situation would be to somehow store just the single bit indicating "we're in an ISR with no _current", and check that everywhere needed.  But that's a ton of auditing.  We might consider moving in that direction on a per-arch basis, e.g. "CONFIG_ARCH_NO_DUMMY_THREAD_NEEDED" or whatever.

[1] Alternatively, we could look at moving these dummies **back** to the interrupt stack, maybe placing them at the bottom.  That doesn't eliminate the cost either but can at least be shared with other uses instead of being a dedicated block.

andyross commented:
Oh, and just to make it clear: this is just a bare `struct _thread`, there's no stack or other auxilliary data to go with it.

npitre commented:
My point is more precisely: why having one dummy for each CPU?
It is a non trivial chunk of memory in some cases. Why not having a 
single one instead? Unless I'm missing something, we don't care if 
multiple CPUs write to it simultaneously.

andyross commented:
Ah, got it.  Mostly just pure conservatism, wanting to be robust vs. architectures that are going to read stuff back out from the thread or use it as scratch space or whatever.  Also having just one would disallow using it as the MP startup dummy because it would be racy if someone wanted to bring up CPUs in parallel.  But I'm quite sure it would work.

npitre commented:
Just ran twister with relevant tests on a single dummy version and the 
world is not falling appart.

ARM64 would certainly benefit from having only one dummy. That's 
about 2KB of wasted memory otherwise.

peter-mitsis commented:
The current version of the is_halting() loop locks/unlocks interrupts. This introduces a subtle hole that your original approach that locked/unlocked sched_spinlock did not have.

1. Let us imagine that CPUx is in that is_halting() loop waiting for CPUy to abort the thread Ty.
2. When CPUy aborts Ty in halt_thread(), one of the first thing it does is to clear the halting bits.
3. Clearing those halting bits breaks CPUx out of its is_halting() loop allowing CPUx to return immediately from z_thread_halt().
4. However, CPUy is not quite done with Ty even though it cleared the halting bits. It still has some cleanup to do in halt_thread().
5. We know that CPUy will be done that cleanup when it releases sched_spinlock.

The net result of the above is that when CPUx (in an ISR) returns from k_thread_abort() in its ISR, there is a short window of time where Ty is still not quite done. This arguably contradicts our documentation which states that "After k_thread_abort() returns, the thread is guaranteed not to be running or to become runnable anywhere on the system." Although it is true that CPUx will need to acquire sched_spinlock as it exits the ISR to determine the next thread to execute, that would still be too late as CPUx could have used other legitimate means to signal a thread running on yet another CPU to do something with Ty's memory.

The original approach that locked/unlocked sched_spinlock  meant that before CPUx could return from z_thread_halt(), it had to first acquire sched_spinlock. This meant that by the time it successfully acquired it, CPUy was guaranteed to be finished all the cleanup on Ty. Thus, when CPUx returned, Ty was as dead as a Monty Python parrot.

andyross committed:
sched: Optimize dummy thread usage on SMP

Nicolas Pitre points out that since these thread structs are just
dummies for the context swtiching, they can be presumed to be "write
only" and thus there's no point in having one per CPU, everyone can
share the same one.

The only gotcha is that we never really documented (nor really have a
place to document) that rule, so it's not theoretically impossible for
an architecture to read back what it might have written underneath
arch_switch().  Leave this in a separate commit for bisection
purposes, but the risk seems very low.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
kernel/sched: Fix lockless ordering in halt_thread()

We've had threads spinning on the thread state bits, but weren't being
careful to ensure that those bits were the last things seen to change
in a halting thread.  Move it to the end, and add a barrier for
correctness.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
OK, it's up.  Hopefully we don't discover anything new.   Left the new patches as separate changes for clarify (it turns out the "spin on is_halting()" was pre-existing anyway from the perspective of a free memory goof, so easy enough to document as a separate bug).

andyross commented:
And of course changing the flag ordering ends up breaking native_sim (which is by definition not order-sensitive as it can't preempt!).  Will have to figure that out..

andyross commented:
Should be cleaned up now; I missed an early return case (no idea why it didn't fail more broadly: seemingly only native_sim and one x86/nokpti arch test were sensitive to state bits on k_thread_suspend?!)

evgeniy-paltsev commented on kernel/sched.c:
```diff
@@ -1382,6 +1389,22 @@ static void halt_thread(struct k_thread *thread, uint8_t new_state)
+		 * code.
+		 */
+		if (dummify && !IS_ENABLED(CONFIG_ARCH_POSIX)) {
+			z_dummy_thread_init(&_thread_dummy);
```
Just to be sure - is it expected that we can set _current on multiple cores to the same _thread_dummy at the same time?

andyross replied:
It's true empirically, yeah.  The only use of that thread pointer is write-only: it's the context save code on interrupt exit.  Your intuition was the same as mine, and originally I had this set up with a per-cpu dummy.  But @npitre points out that involves significant RAM costs, and there's no reason to pay that now if we don't have a failure case.

Obviously if future arch layers do want to play weird tricks with _current, by e.g. wanting to read back stuff they wrote, even though the thread won't/can't ever be scheduled, that's doable by just checking the _THREAD_DUMMY bit in state.

TaiJuWu replied:
I am very curious why we need to init dummy thread here.

andyross replied:
(sorry, wrote this outside the conversation, pasting here to fix) The name of that routine is unfortunate. The only initialization it does is to set the flag bit; it's used to set the _current thread pointer to the dummy, which so far has only happened at CPU initialization time (so that the context switch code doesn't need a special path to enter the "first" thread).

TaiJuWu commented on kernel/spinlock_validate.c:
```diff
@@ -20,10 +20,17 @@ bool z_spin_lock_valid(struct k_spinlock *l)
+		/* Edge case where an ISR aborted _current */
+		return true;
+	}
+	if (tcpu != (_current_cpu->id | (uintptr_t)_current)) {
```
Will this will limit the core number up to 8 for 64 bit machine, 4 cores for 32 bit machine?

andyross replied:
Yes, that's an existing limitation of CONFIG_SPIN_VALIDATE.  It would be easy enough to fix, but the cost would be that enabling spinlock checking would change the size of the struct, which gets packed into all sorts of things.  IMHO that's a bad trade for debug/test features.

andyross commented:
> I am very curious why we need to init dummy thread here.

The name of that routine is unfortunate.  The only initialization it does is to set the flag bit; it's used to set the _current thread pointer to the dummy, which so far has only happened at CPU initialization time (so that the context switch code doesn't need a special path to enter the "first" thread).
